### DISCUSSION TITLE
Possible to monitor win services in general?

### DISCUSSION THREAD

**Discussion by mountaindude** [coreybutler/node-windows#345] [Q&A]
Being able to run Node.js apps as services is awesome, but my current need is to monitor windows services in general.

In particular I want to figure out if a set of services are running or not.
Or even better - be notified when the services are started/stopped.

So, I guess my question is whether node-windows can be set up to monitor and emit messages for any Windows service, or if it will do so only for services that are also installed using node-windows.

**Comment by coreybutler** [comment]
General Windows service monitoring is not a part of this project. Most organizations use Microsoft tools for this, or tools like Nagios.

Node-windows monitors scripts it installs to determine whether it should be restarted, crashed, etc. The wrapper is a lightweight `child_process` connection that keeps the _service_ alive while determining whether the _script_ should be restarted or aborted. If the _script_ should stop or crash, the node-windows wrapper will exit, causing the _service_ to stop. While this is technically monitoring a node script, it's not the same as monitoring a Windows service.

To the best of my knowledge, there aren't any pre-existing tools for monitoring Windows services from Node. There are tools that recreate what Windows does (like pm2), but these operate the same way (i.e. monitoring node child processes).

You could probably create a script to monitor Windows services by polling services, but you'd have to maintain the state of each service to know when the state changes (i.e. stopped to running ot running to stopped), then trigger an event. I do not know if there are any existing modules that do this.

**Comment by mountaindude** [comment]
Thanks for fast reply.
I very much respect the focus of this project so no worries. 

I'll probably wrap the standard `sc` command and poll the state of the services as needed. Will be good enough.
Proper ops tools of course do this very nicely, in this case there is a need to include such monitoring in an existing, stand-alone tool though. Funnily enough, the tool is named "Butler"! Small world.
